runtime/ts/storage/pouchdb/pouch-db-storage.ts

```diff
@@ -0,0 +1,245 @@
+      return provider;
+    }
+    provider.enableReferenceMode();
+    return provider;
+  }
+
+  async _construct(id: string, type: Type, keyFragment: string) {
+    const key = new PouchDbKey(keyFragment);
+    const keystr = key.toString();
+    
+    const provider = this.newProvider(type, undefined, id, key.toString());
+
+    // Used to track changes for the key.
+    this.remoteStateChangedHandlers.set(key.location, provider);
+    
+    return provider;
+  }
+
+  /**
+   * Connect with an existing storage key.  Returns a cached instance if available.
+   * Returns null if no such storage key exists.
+   */
+  async connect(id: string, type: Type, key: string) : Promise<PouchDbStorageProvider> {
+    const imKey = new PouchDbKey(key);
+
+    // TODO(lindner): fail if not created.
+    return this.construct(id, type, key);
+  }
+
+  /** Unit tests should call this in an 'after' block. */
+  shutdown() {
+    // Stop syncing
+    if (PouchDbStorage.syncHandler) {
+      PouchDbStorage.syncHandler.cancel();
+    }
+    // Close databases
+    for (const db of PouchDbStorage.dbLocationToInstance.values()) {
+      db.close();
+    }
+    PouchDbStorage.dbLocationToInstance.clear();
+  }
+
+  /** @inheritDoc */
+  baseStorageKey(type: Type, keyString: string) : string {
+    const key = new PouchDbKey(keyString);
+    key.location = `backingStores/${type.toString()}`;
+    return key.toString();
+  }
+
+  /** @inheritDoc */
+  async baseStorageFor(type: Type, key : string) {
+    if (this.baseStores.has(type)) {
+      return this.baseStores.get(type);
+    }
+    if (this.baseStorePromises.has(type)) {
+      return this.baseStorePromises.get(type);
+    }
+
+    const storagePromise = this._construct(type.toString(), type.collectionOf(), key) as Promise<PouchDbCollection>;
+
+    this.baseStorePromises.set(type, storagePromise);
+    const storage = await storagePromise;
+    assert(storage, 'baseStorageFor should not fail');
+    this.baseStores.set(type, storage);
+    return storage;
+  }
+
+  /** @inheritDoc */
+  parseStringAsKey(s: string) : PouchDbKey {
+    return new PouchDbKey(s);
+  }
+
+  /** Ceates a new Variable or Colleciton given basic parameters */
+  newProvider(type: Type, name, id, key) : PouchDbStorageProvider {
+   if (type.isCollection) {
+      return new PouchDbCollection(type, this, name, id, key);
+    }
+    if (type.isBigCollection) {
+      return new PouchDbBigCollection(type, this, name, id, key);
+    }
+    return new PouchDbVariable(type, this, name, id, key);
+  }
+
+
+  /** Removes everything that a test could have created. */
+  static async resetPouchDbStorageForTesting() {
+    console.log('RESET storage for testing');
+    for (const db of PouchDbStorage.dbLocationToInstance.values()) {
+      await db.allDocs({include_docs: true}).then(allDocs => {
+        return allDocs.rows.map(row => {
+          console.log("- deleting " + row.id + ' ' + row.doc._rev);
```

Comment:
Seems super verbose even for tests?

Reply:
Knowing which keys your test used is informative.  Maybe not the 100th time though, so will remove..